### PR TITLE
Subgraph API change

### PR DESCRIPTION
**Pull request by dellaert**
This is the first PR from some refactoring I did over the break, to get rid of shared pointers for *containers*. It is totally appropriate that we use shared pointers for factors (although, maybe we should not!), but it never made sense (I see now) that we used shared pointers for graphs, as they are not really shuffled around like factors.

The big changes that are coming (branches have not been published yet) are mostly in linear, to get a newer and faster API for elimination -- prompted by development in hybrid branch. This PR removes shared pointers from the subgraph API, for starters.

Adding @jlblancoc as a reviewer, partly to get his opinion on above, and @ProfFan as FYI as I know he has a forked branch with lots of shared pointer changes (presumably moving to std::shared_ptr).

### CODE REVIEW

**Comment by dellaert** [comment]
Some love?

File: gtsam/linear/SubgraphBuilder.cpp [hunk 1]
```diff
@@ -446,30 +446,29 @@ SubgraphBuilder::Weights SubgraphBuilder::weights(
...
-GaussianFactorGraph::shared_ptr buildFactorSubgraph(
-    const GaussianFactorGraph &gfg, const Subgraph &subgraph,
-    const bool clone) {
-  auto subgraphFactors = boost::make_shared<GaussianFactorGraph>();
-  subgraphFactors->reserve(subgraph.size());
+GaussianFactorGraph buildFactorSubgraph(const GaussianFactorGraph &gfg,
+                                        const Subgraph &subgraph,
+                                        const bool clone) {
+  GaussianFactorGraph subgraphFactors;
+  subgraphFactors.reserve(subgraph.size());
```
Comment (varunagrawal) [inline gtsam/linear/SubgraphBuilder.cpp:453]: What's the point of doing a `reserve` if `push_back` below will allocate memory for us?

File: gtsam/linear/linear.i [hunk 2]
```diff
@@ -625,7 +625,7 @@ virtual class SubgraphSolverParameters : gtsam::ConjugateGradientParameters {
 
 virtual class SubgraphSolver  {
   SubgraphSolver(const gtsam::GaussianFactorGraph &A, const gtsam::SubgraphSolverParameters &parameters, const gtsam::Ordering& ordering);
-  SubgraphSolver(const gtsam::GaussianFactorGraph &Ab1, const gtsam::GaussianFactorGraph* Ab2, const gtsam::SubgraphSolverParameters &parameters, const gtsam::Ordering& ordering);
+  SubgraphSolver(const gtsam::GaussianFactorGraph &Ab1, const gtsam::GaussianFactorGraph& Ab2, const gtsam::SubgraphSolverParameters &parameters, const gtsam::Ordering& ordering);
```
Comment (varunagrawal) [inline gtsam/linear/linear.i:628]: Nit: formatting

**Review by varunagrawal — APPROVED** [review]
Looks good to me modulo 2 comments. Would like to hear @jlblancoc's thoughts as well.

File: gtsam/linear/SubgraphBuilder.cpp (hunk 1 above)
Comment (dellaert) [reply gtsam/linear/SubgraphBuilder.cpp:453]: Reserve is recommended if you know exactly how many things you will push

File: gtsam/linear/linear.i (hunk 2 above)
Comment (dellaert) [reply gtsam/linear/linear.i:628]: Global replace…

**Pull request merged**